src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+                                permitsFalse = false;
+                            }
+                            break;
+                        default:
+                            // Once v!=null fails, it must fail a type test
+                            permitsFalse = false;
+                            break;
+                    }
+                    break;
+                case BoundTypeDecision t1:
+                    switch (other)
+                    {
+                        case BoundNonNullDecision n2:
+                            // Once `v is T` is true, v!=null cannot fail
+                            permitsFalse = false;
+                            killsDecisionOnTrueBranch = true;
+                            break;
+                        case BoundTypeDecision t2:
+                            // If T1 could never be T2, then success of T1 implies failure of T2.
+                            var matchPossible = Binder.ExpressionOfTypeMatchesPatternType(Conversions, t1.Type, t2.Type, ref discardedUseSiteDiagnostics, out _);
+                            if (matchPossible == false) permitsTrue = false;
+                            // If every T2 is a T1, then failure of T1 implies failure of T2.
+                            matchPossible = Binder.ExpressionOfTypeMatchesPatternType(Conversions, t2.Type, t1.Type, ref discardedUseSiteDiagnostics, out _);
+                            if (matchPossible == true)
+                            {
+                                // Once we know it is of the subtype, we do not need to test for the supertype.
+                                permitsFalse = false;
+                                killsDecisionOnTrueBranch = true;
+                            }
+                            break;
+                        case BoundValueDecision v2:
+                            break;
+                    }
+                    break;
+                case BoundValueDecision v1:
+                    switch (other)
+                    {
+                        case BoundNonNullDecision n2:
+                            if (v1.Value == ConstantValue.Null)
+                            {
+                                // once v==null is false, we know v!=null is true and do not need to test it
+                                permitsTrue = false;
+                                killsDecisionOnFalseBranch = true;
+                            }
+                            else
+                            {
+                                // once v==K is true, we know v!=null is true and do not need to test it
+                                permitsFalse = false;
+                                killsDecisionOnTrueBranch = true;
+                            }
+                            break;
+                        case BoundTypeDecision t2:
+                            if (v1.Value == ConstantValue.Null) permitsTrue = false;
```

Comment:
>permitsTrue = false; [](start = 64, length = 20)

Please put {} around embedded statement.

Reply:
Addressed in https://github.com/dotnet/roslyn/pull/24059/commits/114a56e9077ed03f08443889b69cae6972a4acc6 of https://github.com/dotnet/roslyn/pull/24059